### PR TITLE
#159206049  Implement email verification of newly created users

### PR DESCRIPTION
**Pull request by klevamane**
#### What does this PR do?
- Send verification email to a newly created user or an unverified user
#### Description of Task to be completed?
- write units test for the response on sending, verifying and reverifying a user
- send a verification email to a new user
- verify generated token
- update user status to verified
#### How should this be manually tested?
Have the following endpoints working

 POST /api/users

 GET /api/users/verify

 GET /api/users/recover

 In postman navigate to `http://localhost:3000/api/users`
  
#### Any background context you want to provide?
- Pull in changes and run `npm install`
- Start server with `npm run dev`

#### What are the relevant pivotal tracker stories?
#159206049

[Delivers #159206049]

### CODE REVIEW

File: config/config.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { } from 'dotenv/config';
+// import { } from 'dotenv/config';
+require('dotenv').config();
```
Comment (veeqtor) [inline config/config.js]: Can we stick with ES6 syntax here?

File: helpers/exports.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  } catch (err) {
+    // probable error: Cannot read property "isunverified" for an email address that doesn't exist
+    // This should be passed to the login
+    winston.info(err);
+  }
+};
+
+exports.checkIfUserIsVerified = async (req, res, next) => {
+  const { email } = req.body.user;
+  // try {
```
Comment (veeqtor) [inline helpers/exports.js]: Seems you left this commented code here, And I feel the nested if statement can be refactored to a single if statement, using the `&&` operator

**Review by veeqtor — APPROVED** [review]

File: controllers/UsersController.js [hunk 3]
```diff
@@ -83,4 +86,40 @@ export default class UsersController {
...
+  // static verifyUser()
+  /**
+/**
+   * Verify the email sent to the newly registered user
+   * @param {*} req - request object
+   * @param {*} res - response object
+   * @param {*} next - Next function
+   * @returns {token} token - returns JWT token
+   */
+  static async verifySentEmail(req, res) {
```
Comment (OKiMaureen) [inline controllers/UsersController.js]: I think a more descriptive name will be verifyEmail. Cause what you are doing is verifying the email Address, not an email sent to them.

File: helpers/exports.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      });
+    }
+    sendVerificationEmail.sendEmail(usertoBeVerified);
+    res.status(200).json({
+      message: {
+        body: ['A verification email will be sent to you if your account is not yet verified']
+      }
+    });
+  } catch (err) {
+    // probable error: Cannot read property "isunverified" for an email address that doesn't exist
```
Comment (OKiMaureen) [inline helpers/exports.js]: I'm not sure you should have comments in this manner,  probably use this  
/**
 * 
 *
*/

Comment (klevamane) [reply helpers/exports.js]: Thank you OkiMaurren. I suppose that's also the recommended way. I'd make the changes right away.

File: controllers/UsersController.js (hunk 3 above)
Comment (klevamane) [reply controllers/UsersController.js]: Thank you @OKiMaureen for this oversight. I Think I get your point. I will act on it.

File: controllers/UsersController.js [hunk 5]
```diff
@@ -64,7 +67,7 @@ export default class UsersController {
           if (result) {
             const token = utils.signToken({ id: foundUser.id });
             res.status(200).json(utils.userToJson(foundUser, token));
-          } else {
+          } else { // password checker
```
Comment (anneKay) [inline controllers/UsersController.js]: why do you have the commented line there?

**Review by OKiMaureen — APPROVED** [review]

File: controllers/UsersController.js [hunk 6]
```diff
@@ -135,8 +138,8 @@ export default class UsersController {
           }, {
             returning: true,
             where: { id: user.id }
-          })
-            .then(([rows, [updatedUser]]) => {
+        })
```
[the remark this answers is not in the document]
Comment (anuonifade) [reply controllers/UsersController.js]: @klevamane you still have a lot of hound ci issues unattended to and I think you increase the coverage.

**Comment by klevamane** [comment]
@andela-aonifade I have corrected the Hound CI issues and fixed the failing test.

File: helpers/exports.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+exports.resendVerificationEmail = async (req, res) => {
+  const { email } = req.body.user;
+  try {
+    const usertoBeVerified = await User.findOne({
+      where: { email }
+    });
+    if (usertoBeVerified === null || usertoBeVerified === 'null' || usertoBeVerified.isverified === true) {
+      return res.status(400).json({
+        errors: {
+          body: ['Your account has already been verified or the email address has not been registered']
```
Comment (anuonifade) [inline helpers/exports.js]: Please separate these two. Account verified should be separated from email registered.

File: helpers/sendmail.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  const token = jwt.sign(
+    { id: userToBeVerified.id },
+    process.env.SECRETE_KEY, { expiresIn: 60 * process.env.VERIFYTOKEN_EXPIRY }
+  );
+
+  // create template link
+  const templateLink = `${process.env.URL_HOST}${token}`;
+  const htmlTempate = mailtemplate.emailTemplate(templateLink);
+  const mailOptions = {
+    from: '"Authors haven" <noreply@authorhaven.sims.com>',
```
Comment (anuonifade) [inline helpers/sendmail.js]: Does this email work? Looks like an email that does not exist.

**Review by anuonifade — CHANGES_REQUESTED** [review]
Please make those changes.

File: helpers/sendmail.js (hunk 8 above)
Comment (klevamane) [reply helpers/sendmail.js]: it's a no reply email. The front end will indicate an email that can be replied to.

Comment (anuonifade) [reply helpers/sendmail.js]: I think that should come from an environment variable for now or fetch it from a database table @klevamane.

**Review by anuonifade — APPROVED** [review]
Ok

**Pull request merged**